crates/iota-grpc-server/src/utils.rs

```
@@ -0,0 +1,84 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+/// Creates a lazy batching stream that fetches and batches items on-demand
+/// based on message size limits.
+///
+/// # Example
+/// ```ignore
+/// create_batching_stream!(
+///     requests.into_iter(),
+///     (object_id, version),
+///     {
+///         let result = process(object_id, version);
+///         let size = result.encoded_len();
+///         (result, size)
+///     },
+///     max_message_size,
+///     GetObjectsResponse,
+///     objects,
+///     has_next
+/// )
+/// ```
+#[macro_export]
+macro_rules! create_batching_stream {
+    (
+        $requests_iter:expr,
+        $item_pattern:pat,
+        $process_block:block,
+        $max_message_size:expr,
+        $response_type:ty,
+        $items_field:ident,
+        $has_next_field:ident
+    ) => {
+        async_stream::try_stream! {
+            let mut requests_iter = $requests_iter;
+            let mut current_batch = Vec::new();
+            let mut current_size = 0;
+            let mut has_yielded = false;
+
+            loop {
+                // Try to get the next item
+                match requests_iter.next() {
+                    Some($item_pattern) => {
+                        // Process the item using the provided block
+                        let (result_item, item_size) = $process_block;
+
+                        // Check if adding this item would exceed the limit
+                        if current_size + item_size > $max_message_size && !current_batch.is_empty() {
```

I think we should return an error if the `current_batch` was empty, but the item still didn't fit.
Otherwise we might exceed the message size in some cases.

---

Thanks for identifying this bug! I fixed it at [85f4747](https://github.com/iotaledger/iota/pull/9328/commits/85f4747c1bf4b7cb663adca2fd36086f5b34f7ae).